measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
 
-        obj = cp.Maximize(objective())
+        ans_y, _ = self.extract_solutions()
+        print("pyomo calculated cost:", pyo.value(self.mod.cost))
+        print("if install dynamic measurements:")
+        print(pyo.value(self.mod.if_install_dynamic[3]))
+
+        if store_name:
+
+            file = open(store_name + str(budget_opt), "wb")
+            pickle.dump(ans_y, file)
+            file.close()
+
+            file2 = open(store_name + "_fim_" + str(budget_opt), "wb")
+            pickle.dump(fim_result, file2)
+            file2.close()
+
+    def _locate_initial_file(self, budget):
+        """
+        Given the budget, select which initial solution it should use by locating the closest budget that has stored solutions
+        It selects according to:
+        1) if this budget has a previous solution, find it, return the file name
+        2) if this budget does not have a previous solution, find the closest budget that has a solution, return the file name
+
+        Arguments
+        --------
+        budget: budget
+
+        Return
+        ------
+        y_init_file: the file name that contains a previous solution
+        """
+        ## find if there has been a original solution for the current budget
+        if budget in self.curr_res_list:  # use an existed initial solutioon
+            y_init_file = self.curr_res_list[budget]
+            curr_budget = budget
+
+        else:
+            # if not, find the closest budget, and use this as the initial point
+            curr_min_diff = np.inf  # current minimal budget difference
+            curr_budget = max(list(self.curr_res_list.keys()))  # starting point
+
+            # find the existing budget that minimize curr_min_diff
+            for i in list(self.curr_res_list.keys()):
+                # if we found an existing budget that is closer to the given budget
+                if abs(i - budget) < curr_min_diff:
+                    curr_min_diff = abs(i - budget)
+                    curr_budget = i
+
+        if self.precompute_print_level >= 1:
+            print("using solution at", curr_budget, " too initialize")
+
+        # assign solution file names, and FIM file names
+        y_init_file = self.curr_res_list[curr_budget]
+
+        return y_init_file
+
+    def _initialize_binary(self, y_init_file):
+        """This function initializes all binary variables directly in the created model
+
+        Arguments
+        ---------
+        y_init_file: the file name that contains a previous solution
+
+        Return
+        ------
+        None. self.mod is updated with initial values for all binary decisions
+        """
+        # read y
+        with open(y_init_file, "rb") as f:
+            init_cov_y = pickle.load(f)
+
+        # Round possible float solution to be integer
+        for i in range(self.num_measure_dynamic_flatten):
+            for j in range(self.num_measure_dynamic_flatten):
+                if init_cov_y[i][j] > 0.99:
```

Comment:
Any hard-coded values list should be declared and set in the object's __init__ method. This tells future programmers about possible tuning parameters for the algorithm

Reply:
- [ ] change this to init, do not need people to change this.

Reply:
added them as `near_1/0_threshold` to init method.